src/QuadForm/Quad/ZLattices.jl

```diff
@@ -1813,3 +1813,82 @@ function overlattice(glue_map::TorQuadModMor)
   return lattice(ambient_space(S), B[end-rank(S)-rank(R)+1:end,:])
 end
 
+################################################################################
+#
+#  Primary/elementary lattices
+#
+################################################################################
+
+@doc Markdown.doc"""
+    is_primary_with_prime(L::ZLat) -> Bool, fmpz
+
+Given a $\mathbb Z$-lattice `L`, return whether `L` is primary, that is whether its
+discriminant group is a `p`-group for some prime number `p`. In case it is, `p` is
+also returned as second output.
+
+Note that for unimodular lattices, this function returns `(true, 1)`. If the
+lattice is not primary, the second return value is `-1` by default.
+"""
+function is_primary_with_prime(L::ZLat)
+  @req is_integral(L) "L must be integral"
+  d = ZZ(abs(det(L)))
+  if d == 1
+    return true, d
+  end
+  pd = prime_divisors(d)
+  if length(pd) != 1
+    return false, ZZ(-1)
+  end
+  return true, pd[1]
+end
+
+@doc Markdown.doc"""
+    is_primary(L::ZLat, p::Union{Integer, fmpz}) -> Bool
+
+Given a $\mathbb Z$-lattice `L` and a prime number `p`, return whether `L` is
+`p`-primary, that is whether its discriminant group is a `p`-group.
+"""
+function is_primary(L::ZLat, p::Union{Integer, fmpz})
+  bool, q = is_primary_with_prime(L)
+  return bool && q == p
+end
+
+@doc Markdown.doc"""
+    is_unimodular(L::ZLat) -> Bool
+
+Given a $\mathbb Z$-lattice `L`, return whether `L` is unimodular, that is
+whether its discriminant group is trivial.
+"""
+is_unimodular(L::ZLat) = is_primary(L, 1)
+
+@doc Markdown.doc"""
+    is_elementary_with_prime(L::ZLat) -> Bool, fmpz
+
+Given a $\mathbb Z$-lattice `L`, return whether `L` is elementary, that is whether
+its discriminant group is an elemenentary `p`-group for some prime number `p`. In
+case it is, `p` is also returned as second output.
+
+Note that for unimodular lattices, this function returns `(true, 1)`. If the lattice
+is not elementary, the second return value is `-1` by default.
+"""
+function is_elementary_with_prime(L::ZLat)
+  bool, p = is_primary_with_prime(L)
+  bool || return false, ZZ(-1)
+  if !is_integer(p*scale(dual(L)))
+    return false, ZZ(-1)
+  end
+  return bool, p
+end
+
+@doc Markdown.doc"""
+    is_elementary(L::ZLat, p::Union{Integer, fmpz}) -> Bool
+
+Given a $\mathbb Z$-lattice `L` and a prime number `p`, return whether `L`
+is `p`-elementary, that is whether its discriminant group is an elementary
+`p`-group.
+"""
+function is_elementary(L::ZLat, p::Union{Integer, fmpz})
```

Comment:
Is this standard definition? Since L itself is an abelian group, it could be elementary itself (which it never will be, but still). It is like defining the order of L to be the order of the discriminant group.

But I am not too familiar with the literature. @simonbrandhorst any thoughts?

Reply:
I am following the definition of Conway and Sloane for p-elementary lattices yes. For primary I am not sure whether it is standard, but it is coherent with the elementary definition... I remember seeing it used in some paper also.

Reply:
Ok. It just looks odd to forward some properties of L to the discrimant group (and some not). But I am happy with it if this the convention.

Reply:
Good. Simon wanted to have a look at it in any case so he will be able to confirm (or not!)

Reply:
Yes, `p`-elementary for `p` a prime is standard.